common/workunit/workunit.cpp

```diff
@@ -332,113 +332,1203 @@ extern WORKUNIT_API IConstWUGraphProgress *createConstGraphProgress(const char *
+    }
+    virtual StatisticScopeType getScopeType() const
+    {
+        return scopeType;
+    }
+    virtual unsigned __int64 getValue() const
+    {
+        return value;
+    }
+    virtual unsigned __int64 getCount() const
+    {
+        return count;
+    }
+    virtual unsigned __int64 getMax() const
+    {
+        return max;
+    }
+    virtual unsigned __int64 getTimestamp() const
+    {
+        return timeStamp;
+    }
+    virtual bool matches(const IStatisticsFilter * filter) const
+    {
+        return filter->matches(creatorType, creator, scopeType, scope, measure, kind, value);
+    }
+
+public:
+    StringBuffer creator;
+    StringBuffer description;
+    StringBuffer scope;
+    StatisticMeasure measure;
+    StatisticKind kind;
+    StatisticCreatorType creatorType;
+    StatisticScopeType scopeType;
+    unsigned __int64 value;
+    unsigned __int64 count;
+    unsigned __int64 max;
+    unsigned __int64 timeStamp;
+};
+
+//---------------------------------------------------------------------------------------------------------------------
+
+/*
+ * The following compare functions are used to ensure that comparison are consistent with
+ * compareScopeName() in jstats.  This ensures that the scope iterators from different sources
+ * are processed in a consistent order
+ */
+
+static int compareGraphNode(IInterface * const *ll, IInterface * const *rr)
+{
+    IPropertyTree *l = (IPropertyTree *) *ll;
+    IPropertyTree *r = (IPropertyTree *) *rr;
+    const char * lname = l->queryName();
+    const char * rname = r->queryName();
+    return compareScopeName(lname, rname);
+}
+
+static int compareSubGraphStatsNode(IInterface * const *ll, IInterface * const *rr)
+{
+    IPropertyTree *l = (IPropertyTree *) *ll;
+    IPropertyTree *r = (IPropertyTree *) *rr;
+    return compareScopeName(l->queryName(), r->queryName());
+}
+
+static int compareSubGraphNode(IInterface * const *ll, IInterface * const *rr)
+{
+    IPropertyTree *l = (IPropertyTree *) *ll;
+    IPropertyTree *r = (IPropertyTree *) *rr;
+    return l->getPropInt("@id") - r->getPropInt("@id");
+}
+
+static int compareActivityNode(IInterface * const *ll, IInterface * const *rr)
+{
+    IPropertyTree *l = (IPropertyTree *) *ll;
+    IPropertyTree *r = (IPropertyTree *) *rr;
+    return l->getPropInt("@id") - r->getPropInt("@id");
+}
+
+static int compareEdgeNode(IInterface * const *ll, IInterface * const *rr)
+{
+    IPropertyTree *l = (IPropertyTree *) *ll;
+    IPropertyTree *r = (IPropertyTree *) *rr;
+    //MORE: Edge needs more work
+    const char * leftId = l->queryProp("@id");
+    const char * rightId = r->queryProp("@id");
+    unsigned leftAc = atoi(leftId);
+    unsigned rightAc = atoi(rightId);
+    if (leftAc != rightAc)
+        return (int)(leftAc - rightAc);
+    const char * leftSep = strchr(leftId, '_');
+    const char * rightSep = strchr(rightId, '_');
+    assertex(leftSep && rightSep);
+    return atoi(leftSep+1) - atoi(rightSep+1);
```

Comment:
Correct - there needs to be a well defined order that all sources can iterate efficiently.  Which requires it to be numeric not alphabetic (because of graphs)